### ISSUE TITLE
Doesn't work correctly with overlapping drop targets

### ISSUE THREAD

**Issue by gnapse** [yahoo/react-dnd-touch-backend#14]
I have a trello-like interface in place, and it works great with the html5 backend. I decided to switch to this backend because I want to support touch devices, and also I'm planning to provide a custom drag layer anyway, so I should be able to ditch html5 backend completely, and use this backend with mouse events support enabled, to cover both desktop and touch devices.

Anyway, when I switched, it "works", but since both my list columns and the cards within them are drop targets, this seems to cause a problem with this backend. The behavior I'm noticing is consistent with it ignoring the cards as drop target altogether, and always considering that the drop target is the list containing the card.

The reason I think this is what's happening, is that when I drag a card from one list to another, it does not matter if I'm hovering a card in the middle of the target list, the drop is made at the end of the list, which is the expected behavior when I'm hovering over an empty area of the list element. If I hover over a card in the middle, the expected behavior is to mark the drop in that place, shifting the cards below one position down.

I may be able to extract the code into a simpler standalone app for sharing and testing, but right now I only have it as part of a more complex application with backend and all, which is not suitable for making it available for sharing/testing. But I decided to raise the issue anyway in case anyone has any ideas of why this is happening.

**Comment by longlho** [comment]
The mouse support is mainly for hybrid devices and not very well-tested. In the mean time I suggest you use both `HTML5Backend` & `TouchBackend` conditionally while I'm tryin to take a look. And an example for your use case would be great :)

**Comment by gnapse** [comment]
Ok I'll work on making a standalone app as example. And thanks for the tip on the hybrid thing, although the behavior I describe occurs consistently on both touch and mouse devices.

**Comment by gnapse** [comment]
Ok, I think I narrowed it down to the fact that `monitor.isOver({shallow: true})` is not working correctly.

In my `hover` handler on the list component, there are two possible item types to handle: 'list' or 'card'. When hovering a card over a list, I check if `monitor.isOver({shallow: true})`, in which case I assume the card is being dragged directly over an empty area of the list. Otherwise it means that it is being dragged over another card already within that list, so that is taken care by the `hover` handler on the card component.

But since `monitor.isOver({shallow: true})` is always returning true no matter what, then the behavior is consistent with what should happen when hovering over empty areas of the list component. It should return `false` if hovering over a nested drop target, as is documented [here](http://gaearon.github.io/react-dnd/docs-drop-target.html#specification-methods).

**Comment by longlho** [comment]
hmm I don't think the backend is handling `isOver` implementation thou.

**Comment by gnapse** [comment]
Yeah, I know. I did some research on the ReactDnD and backends source code and it seems so. It's actually on the `dnd-core` package where this is handled. I'll build a sample app where the issue can be replicated, it might take me a few days though.

**Comment by gnapse** [comment]
Here's the example extracted to a standalone app: https://github.com/gnapse/react-dnd-trello

It behaves as described above. There are two commits in that repo. On the initial commit the app uses the Html5 backend, and works as expected. You can checkout that commit and see how it works. Then if you checkout master once again, you can see it working with the touch backend. When you drag a card out of a list, and without dropping it you drag it back into that list, it will always mark the drop at the end, instead of in the middle, no matter where you position it in the list of cards.

**Comment by gnapse** [comment]
@longlho have you tried the sample app? any ideas?

**Comment by longlho** [comment]
sry I haven't got time :(

**Comment by Dennisimmo** [comment]
Hey friends, ive been running into the same issue with my nested droptargets when changing to touchbackend.
I further investigated and found that the problem is related to the order of `matchingTargetIds` and `this.targetNodes` in `handleTopMoveCapture`. This is because the dnd core uses the index of these targetIds to check shallowness in DragAndDropMonitor.js .
 ` if (shallow) {
      return index === targetIds.length - 1;
    } else {
      return index > -1;
    }
`
While the htmlBackend somehow handles to keep the order, touchbackend does not keep the order. In detail: when starting to drag an object, the order of `matchingTargetIds` is right, but as soon as the initial droptarget is left, the order of `matchingTargetIds` is reversed.
For my own app i changed 
`  this.actions.hover(matchingTargetIds, {
            clientOffset: clientOffset
        });` 
to
`  this.actions.hover(matchingTargetIds.reverse(), {
            clientOffset: clientOffset
        });` in `touch.js`
which leads to the expected behaviour of `monitor.isOver({shallow: true})`, but only after changing the droptarget once.
Sadly i did not find the right place to completely fix the problem. Good luck.

**Comment by gnapse** [comment]
Great findings @Dennisimmo!! Thanks a lot. I believe this is a huge step forward in finding the solution. I doubt I'll have significant time anytime soon, considering that I'm not fully familiar with the internals of this backend lib, but I'll definitely be taking a look at it at some point if no one else beats me to it before that. These findings at least give me a clue. Thanks again.

**Comment by gnapse** [comment]
I can confirm @Dennisimmo's findings.

I got a copy of the UMD release of this touch backend into my project, and manually modified [this line](https://github.com/yahoo/react-dnd-touch-backend/blob/e9820c273e1e81aab79f7acbca1e58ae2001f6e3/src/Touch.js#L231) according to his instructions (`matchingTargetIds.reverse()` instead of just `matchingTargetIds`) and my example now works as expected.

Can anyone confirm that this change can be made as is, without breaking anything else?

**Comment by longlho** [comment]
thanks @Dennisimmo :) yeah throw in a PR and add in your use case in the `examples` & that should be good.

**Comment by gnapse** [comment]
I now understand fully the comment by @Dennisimmo above, and why his fix is not fully a correct fix:

> when starting to drag an object, the order of `matchingTargetIds` is right, but as soon as the initial droptarget is left, the order of `matchingTargetIds` is reversed.

At least in my example it is not necessary to leave the initial drop target in order for the fix to work. But still during an initial moment after the drag starts, I notice a flickering in which, upon starting to drag, the recently "lifted" card briefly appears as droppable in the end of the list instead of on the very own spot from which it was recently grabbed. Then the drop mark goes back to where it is supposed, but the flickering is very obvious.

Here's a short video of what I just described: http://cl.ly/13241e2Y0h0m (I haven't implemented a drag layer so that's why you don't see a "ghost" accompanying the mouse cursor). This video is with the above fix in place, but it clearly still has a problem in the initial moment of the drag.

**Comment by longlho** [comment]
hmm kk im taking a look now. The order of `matchingTargetIds` is non-deterministic conceptually (since it's a filter of `key`s on an `object`) so the `reverse` fix won't work for all cases.

**Comment by longlho** [comment]
@gnapse just put a PR in your example

**Comment by glenn-allen** [comment]
Hi Guys, just wondering if anyone has made any progress on this?
My team has encountered the same issue, so rather than reinvent the wheel I thought I'd check.
Thanks

**Comment by longlho** [comment]
I put in a [PR](https://github.com/gnapse/react-dnd-trello/pull/1) for @gnapse repo but haven't heard back

**Comment by gnapse** [comment]
Sorry, I recall having checked that at the time, and I recall it fixed the problem. Thanks @longlho. So @glenn-allen perhaps you can apply the corresponding fix in your app to see if it works, because the mentioned PR is not on this library, but on the app that uses it. I guess this issue can be closed now, but I'll take the time to double check now that fix and will come back here with the results later these days.

**Comment by longlho** [comment]
@gnapse feel free to reopen :)

**Comment by glenn-allen** [comment]
OK, thanks for the suggestions.
I don't think that will help our case because we can have overlapping drop targets of various types (as it's a free form canvas), so we really need the `monitor.isOver({shallow: true})` to only return true when the element at the highest stacking context is hovered.
From my initial searches this doesn't seem to be an easy problem to solve, if I get the chance to delve deeper over the next few days I'll put together a demo. 
Either way thanks for the amazing work on the library, it really is greatly appreciated!

**Comment by glenn-allen** [comment]
I gather the cause issue is known but the fix perhaps isn't so I've created a repo that demonstrates the issue and might help with the investigation (extended from the react dnd example).
[https://github.com/glenn-allen/dnd-touch-example](https://github.com/glenn-allen/dnd-touch-example)

Let me know if I should open a separate issue for this.
